{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": []
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "code",
      "metadata": {
        "id": "l0IZC108RJjx"
      },
      "source": [
        "# parse a fasta file\n",
        "def parse_fasta_file(input_file):\n",
        "    parsed_seqs = {}\n",
        "    curr_id = None\n",
        "    curr_seq = []\n",
        "\n",
        "    for line in input_file:\n",
        "        line = line.strip()\n",
        "\n",
        "        if line.startswith('>'):\n",
        "            if curr_id is not None:\n",
        "                parsed_seqs[curr_id] = ''.join(curr_seq)\n",
        "\n",
        "            curr_id = line[1:]\n",
        "            curr_seq = []\n",
        "            continue\n",
        "\n",
        "        curr_seq.append(line)\n",
        "\n",
        "    # add final sequence to dict\n",
        "    parsed_seqs[curr_id] = ''.join(curr_seq)\n",
        "    return parsed_seqs\n",
        "\n",
        "#------------------------------------------------------------------------------\n",
        "# parse a fasta file and replace ids\n",
        "def parse_fasta_file2(input_file, keys):\n",
        "    parsed_seqs = {}\n",
        "    curr_id = None\n",
        "    curr_seq = []\n",
        "    i = 0\n",
        "\n",
        "    for line in input_file:\n",
        "        line = line.strip()\n",
        "\n",
        "        if line.startswith('>'):\n",
        "            if curr_id is not None:\n",
        "                parsed_seqs[curr_id] = ''.join(curr_seq)\n",
        "\n",
        "            curr_id = keys[i]\n",
        "            curr_seq = []\n",
        "            i+=1\n",
        "            continue\n",
        "\n",
        "        curr_seq.append(line)\n",
        "\n",
        "    # add final sequence to dict\n",
        "    parsed_seqs[curr_id] = ''.join(curr_seq)\n",
        "    return parsed_seqs\n",
        "\n",
        "#------------------------------------------------------------------------------\n",
        "# print a fasta file\n",
        "def print_fasta_file(file_name):\n",
        "#     infile = open(file_name)\n",
        "    for line in infile:\n",
        "        print(line.strip())\n",
        "\n",
        "#------------------------------------------------------------------------------\n",
        "# write a fasta file\n",
        "def write_fasta_file(output_path, sequences, line_chars=80):\n",
        "    output_file = open(output_path, 'w')\n",
        "\n",
        "    for seq_id, seq in sequences.items():\n",
        "        id_line = '>' + seq_id + \"\\n\"\n",
        "        output_file.write(id_line)\n",
        "\n",
        "        if len(seq) < line_chars:\n",
        "            sequence_line = seq + \"\\n\"\n",
        "            output_file.write(sequence_line)\n",
        "            continue\n",
        "\n",
        "        seq_lines = [seq[i:i+line_chars] for i in range(0,len(seq),line_chars)]\n",
        "        for seq_line in seq_lines:\n",
        "            output_file.write(seq_line + \"\\n\")\n",
        "\n",
        "    # close file when done\n",
        "    output_file.close()"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "geLq8_Ak_fKk"
      },
      "source": [
        "#Parse aligned sequence file and replace sequence ids with variant names\n",
        "input_file = open(\"covid_variants_output1.fa\")\n",
        "seq_names = ['Wuhan','Alpha','Beta','Gamma','Delta','Kappa','Eta','Iota','Lambda']\n",
        "time = {'Wuhan':0,'Alpha':38,'Beta':87,'Gamma':98,'Delta':266,'Kappa':63,'Eta':346,'Iota':112,'Lambda':313}\n",
        "seqs = parse_fasta_file2(input_file, seq_names)\n",
        "\n",
        "# print(parsed_seqs)\n",
        "\n",
        "print(\"Sequence lengths:\")\n",
        "for seq in seqs:\n",
        "  print(f\"{seq}:\",len(seqs[seq]))"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "TQ4BGq5pWPs5"
      },
      "source": [
        "#------------------------------------------------------------------------------\n",
        "# Calculates the Mutation Rate between two different genomic sequences\n",
        "def MutationRate(positions, genome1, genome2):\n",
        "    mutationrate = 0\n",
        "    if len(genome1) == len(genome2):\n",
        "        mutationrate = len(positions) / (2 * len(genome1))\n",
        "    if len(genome1) > len(genome2):\n",
        "        mutationrate = len(positions) / len(genome2)\n",
        "    if len(genome1) < len(genome2):\n",
        "        mutationrate = len(positions) / len(genome1)\n",
        "    return mutationrate\n",
        "\n",
        "#------------------------------------------------------------------------------\n",
        "#Comparison Function\n",
        "#Iterates through a genomic sequence and returns the indexes where differences\n",
        "#occur\n",
        "def Comparison(genome1, genome2):\n",
        "    positions = []\n",
        "    if (len(genome1) == len(genome2)):\n",
        "        for i in range(len(genome1)):\n",
        "            if genome1[i] != genome2[i]:\n",
        "                positions.append(i)\n",
        "        return positions\n",
        "\n",
        "    if (len(genome1) > len(genome2)):\n",
        "        for i in range(len(genome2)):\n",
        "            if genome1[i] != genome2[i]:\n",
        "                positions.append(i)\n",
        "        if len(positions) == 0:\n",
        "            positions.append(len(genome1) - 1)\n",
        "        return positions\n",
        "    if (len(genome1) < len(genome2)):\n",
        "        for i in range(len(genome1)):\n",
        "            if (genome1[i] != genome2[i]):\n",
        "                positions.append(i)\n",
        "        if len(positions) == 0:\n",
        "            positions.append(len(genome2) - 1)\n",
        "        return positions\n",
        "\n",
        "#hi"
      ],
      "execution_count": null,
      "outputs": []
    }
  ]
}